x-pack/plugins/siem/common/endpoint/generate_data.ts

```diff
@@ -319,6 +364,78 @@ export class EndpointDocGenerator {
     };
   }
 
+  /**
+   * This generates a full resolver tree and keeps the entire tree in memory. This is useful for tests that want
+   * to compare results from elasticsearch with the actual events created by this generator. Because all the events
+   * are stored in memory do not use this function to generate large trees.
+   *
+   * @param options - options for the layout of the tree, like how many children, generations, and ancestry
+   * @returns a Tree structure that makes accessing specific events easier
+   */
+  public generateTree(options: TreeOptions = {}): Tree {
+    const addEventToMap = (nodeMap: Map<string, TreeNode>, event: Event) => {
+      const nodeId = event.process.entity_id;
+      // if a node already exists for the entity_id we'll use that one, otherwise let's create a new empty node
+      // and add the event to the right array.
+      let node = nodeMap.get(nodeId);
+      if (!node) {
+        node = { id: nodeId, lifecycle: [], relatedEvents: [] };
+      }
+
+      // place the event in the right array depending on its category
+      if (event.event.category === 'process') {
+        node.lifecycle.push(event);
+      } else {
+        node.relatedEvents.push(event);
+      }
+      return nodeMap.set(nodeId, node);
+    };
+
+    const ancestry = this.createAlertEventAncestry(
+      options.ancestors,
+      options.relatedEvents,
+      options.percentWithRelated,
+      options.percentTerminated
+    );
+
+    // create a mapping of entity_id -> lifecycle and related events
+    const ancestryNodes: Map<string, TreeNode> = _.reduce(
```

Comment:
Good point, I'll switch it to use the array reduce.

Reply:
@oatkiller It's slower and worse (in 2020 - was actually not the case in a lot of engines 4ish years ago - lodash gave better perf on some things by doing some weird stuff under the hood). Doing it that way was my suggestion because it was a _.groupBy before and I figured one lodash collection function for another was easy (there is some neat stuff you can do like chaining and tapping that stuff, too, but that doesn't come into play here.)

Reply:
I've avoided lodash for the past couple of years because the reasons to use it have attenuated somewhat (compared to, say, 2014ish) and I know you and @kqualters-elastic don't care for it, but in its defense it has some pretty decent facilities for doing a lot of stuff to arrays and objects in fairly terse and readable ways. I'm not itching to start using it again - I generally agree with the reasons I think we tend to avoid it as a team, but it's not necessarily all bad.

Reply:
e.g. you could take out the function above _and_ the whole addToEventMap function and replace it with:

`````js
var lodashrules = _(events)
  .groupBy('process.entity_id')
  .map((evts,entity_id)=>{
    const [lifecycle, relatedEvents] = _.partition(evts,e=>e.process.category==='process')
    return {id: entity_id, lifecycle, relatedEvents}
  })
  .reduce((acc,val)=>{ return acc.set(val.id, val) }, new Map())
`````

I'm _not_ saying we should do that, just pointing out that sometimes lodash is pretty cool.